tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java

```diff
@@ -164,357 +190,371 @@ public void invalidateBufferedData() throws IOException {
         this.write(b, 0, b.length);
     }
-    
-    /* overrides RandomAccessFile.write(byte[], int, int) */
+
+    @Override
     public void write(byte[] b, int off, int len) throws IOException {
         // Assert.check(!this.closed);
+        assert super.getFilePointer() == diskPos;
         while (len > 0) {
             int n = this.writeAtMost(b, off, len);
             off += n;
             len -= n;
+            assert super.getFilePointer() == diskPos;
         }
-        this.dirty = true;
     }
 
-    public void writeBigInteger(BigInteger bi, int size) throws IOException {
-      byte[] b = bi.toByteArray();
-      // Assert.check(b.length <= size);
-      this.write(b, 0, size);
-    }
-    
     /* Precondition: x is a non-negative short. */
     public final void writeShortNat(int x) throws IOException {
-      if (x <= 0x7f) {
-        this.writeByte((short)x);
-      }
-      else {
-        this.writeShort(-x);
-      }
+        if (x <= 0x7f) {
+            this.writeByte((short)x);
+        } else {
+            this.writeShort(-x);
+        }
     }
 
-  /* Precondition: x is a non-negative int. */
-  public final void writeNat(int x) throws IOException {
-    if (x <= 0x7fff) {
-      this.writeShort((short)x);
-    }
-    else {
-      this.writeInt(-x);
+    /* Precondition: x is a non-negative int. */
+    public final void writeNat(int x) throws IOException {
+        if (x <= 0x7fff) {
+            this.writeShort((short)x);
+        } else {
+            this.writeInt(-x);
+        }
     }
-  }
 
-  /* Precondition: x is a non-negative long. */
-  public final void writeLongNat(long x) throws IOException {
-    if (x <= 0x7fffffff) {
-      this.writeInt((int)x);
-    }
-    else {
-      this.writeLong(-x);
+    /* Precondition: x is a non-negative long. */
+    public final void writeLongNat(long x) throws IOException {
+        if (x <= 0x7fffffff) {
+            this.writeInt((int)x);
+        } else {
+            this.writeLong(-x);
+        }
     }
-  }
-
-    /* Write at most "len" bytes to "b" starting at position "off",
-       and return the number of bytes written. */
-    private int writeAtMost(byte[] b, int off, int len)
-      throws IOException {
-        if (this.curr == this.hi) {
-            if (this.hitEOF && this.hi < this.maxHi) {
-                // at EOF -- bump "hi"
-                this.hi = this.maxHi;
-            } else {
-                // slow path -- write current buffer; read next one
-                this.seek(this.curr);
-                if (this.curr == this.hi) {
-                    // appending to EOF -- bump "hi"
-                    // Assert.check(this.hitEOF);
-                    this.hi = this.maxHi;
-                }
-            }
+
+    /**
+     * Write at most "len" bytes from "b" starting at position "off", and return the number of bytes written.
+     */
+    private int writeAtMost(byte[] b, int off, int len) throws IOException {
```

Comment:
It seems like it may be worth greater optimization here, given that, in theory, that is the motivation for this class.  This implementation makes sense, in that it is quite safe.  If I understand it, we are simply pushing data into the buffer, up to what it can handle, and if we filled the buffer, we use seek to flush the current buffer and move it upwards a single frame, please correct me if I've misunderstood anything!

However, if we write a substantial amount of data, we're a bit wasteful on two fronts.  For one, we copy data into the buffer, which will never be read, when we could directly write the data we have.  For the second, we are going to fill the buffer with data when the frame moves, and do a read from disk for data that we will never read and immediately overwrite.

It seems like we should detect these cases when we're doing a long enough write and bypass the longer route, leaving the buffer in a temporarily invalid state that we know we'll clean up at the very last frame move.  Otherwise, it seems like this class may not be living up to its mission of optimizing file interactions.

Reply:
I thought about this as well---but let's defer that to future optimizations. I will add a note about this idea.

FWIW, TLC's workload on this class does not involve large bulk writes. It reads and writes almost entirely single-byte and 4-byte chunks. But, I believe it _does_ alternate between read phases and write phases, so there may be some unnecessary reads during the write phase. Even so, in my microbenchmarks the buffer more than pays for the extra reads.

Reply:
On the topic of future optimizations (not for this PR), it did strike me that making this a ring buffer over an arbitrary position might help avoid cases where the buffer position flaps back and forth between two frames, causing a series of reads and flushes.  It also might better work with an L2 cache by keeping buffer positions more stable in memory for longer.

Though, again, I’d be curious your estimation of how such changes might actually impact the TLC specific workload.